### DISCUSSION TITLE
How/where to get started with no domain knowledge?

### DISCUSSION THREAD

**Discussion by guccigang420** [NVIDIA/open-gpu-kernel-modules#418] [Q&A]
Hi,

I am a programmer who would like to understand and contribute to this project. However, I have never programmed kernel modules, drivers, or have an idea about the GPU architecture and APIs.

I currently do not have an idea of what this kernel module really does. I assume, that as a driver, any kind of communication with the GPU is done through this module. I also read the brief documentation on the README regarding the separation of os-agnostic and kernel-specific code.

Would anyone recommend some starting points to read, or some documentation on the hardware this interacts with? Or any kind of basic explanation would also be appreciated regarding the high-level functions of this module.

Thank you for your attention.

**Comment by aaronp24** [comment]
Are you familiar with device drivers and how they typically interact with hardware in general? If not, I'd recommend reading something like _Linux Kernel Development_ by Robert Love and taking a look at how a driver for a simpler device such as a serial port works before diving into GPU driver development. GPUs are extremely complex devices with multiple embedded microcontrollers, DMA engines, and internal processing engines and interacting with them is a complex dance between the kernel driver, the user-mode drivers, and the hardware.
